src/stratis_cli/_actions/_pool.py

```diff
@@ -0,0 +1,664 @@
+        pools_to_blockdevs[pool_map[modev.Pool()]].append(str(modev.Devnode()))
+
+    return dict(
+        (pool, frozenset(blockdevs)) for pool, blockdevs in pools_to_blockdevs.items()
+    )
+
+
+def _check_opposite_tier(managed_objects, to_be_added, other_tier):
+    """
+    Check whether specified blockdevs are already in the other tier.
+
+    :param managed_objects: the result of a GetManagedObjects call
+    :type managed_objects: dict of str * dict
+    :param to_be_added: the blockdevs to be added
+    :type to_be_added: frozenset of str
+    :param other_tier: the other tier, not the one requested
+    :type other_tier: _stratisd_constants.BlockDevTiers
+    :raises StratisCliInUseOtherTierError: if blockdevs are used by other tier
+    """
+    pools_to_blockdevs = _generate_pools_to_blockdevs(
+        managed_objects, to_be_added, other_tier
+    )
+
+    if pools_to_blockdevs != {}:
+        raise StratisCliInUseOtherTierError(
+            pools_to_blockdevs,
+            BlockDevTiers.Data
+            if other_tier == BlockDevTiers.Cache
+            else BlockDevTiers.Cache,
+        )
+
+
+def _check_same_tier(pool_name, managed_objects, to_be_added, this_tier):
+    """
+    Check whether specified blockdevs are already in the tier to which they
+    are to be added.
+
+    :param managed_objects: the result of a GetManagedObjects call
+    :type managed_objects: dict of str * dict
+    :param to_be_added: the blockdevs to be added
+    :type to_be_added: frozenset of str
+    :param this_tier: the tier requested
+    :type this_tier: _stratisd_constants.BlockDevTiers
+    :raises StratisCliPartialChangeError: if blockdevs are used by this tier
+    :raises StratisCliInUseSameTierError: if blockdevs are used by this tier in another pool
+    """
+    pools_to_blockdevs = _generate_pools_to_blockdevs(
+        managed_objects, to_be_added, this_tier
+    )
+
+    owned_by_current_pool = frozenset(pools_to_blockdevs.get(pool_name, []))
+    owned_by_other_pools = dict(
+        (pool, devnodes)
+        for pool, devnodes in pools_to_blockdevs.items()
+        if pool_name != pool
+    )
+
+    if owned_by_current_pool != frozenset():
+        raise StratisCliPartialChangeError(
+            "add to cache" if this_tier == BlockDevTiers.Cache else "add to data",
+            to_be_added.difference(owned_by_current_pool),
+            to_be_added.intersection(owned_by_current_pool),
+        )
+    if owned_by_other_pools != {}:
+        raise StratisCliInUseSameTierError(owned_by_other_pools, this_tier)
+
+
+def _fetch_locked_pools_property(proxy):
+    """
+    Fetch the LockedPools property from stratisd.
+    :param proxy: proxy to the top object in stratisd
+    :return: list of pool UUIDs as strings
+    :rtype: list of str
+    :raises StratisCliPropertyNotFoundError:
+    :raises StratisCliEnginePropertyError:
+    """
+    return _fetch_property(proxy, "LockedPools")
+
+
+def _fetch_property(proxy, property_name):
```

Comment:
I _think_ this is the only method that's shared between the two files. Best place for it is ```_action/_utils.py```. There's another function in ```_utils.py``` called ```fetch_property``` already, you can rename that one to something more appropriate, like maybe ```unpack_property```.

Reply:
@mulkieran moving `fetch_property` into `_action/_utils.py` creates a circular import. `_action/_data.py` imports `get_timeout` from `_action/_utils.py`, and `_action/_utils.py` imports `FetchProperties` from `_action/_data.py`. 

How do you feel about moving `get_timeout` to `_action/_data.py`, as that is the only place outside of testing where it is used?

Reply:
> @mulkieran moving `fetch_property` into `_action/_utils.py` creates a circular import. `_action/_data.py` imports `get_timeout` from `_action/_utils.py`, and `_action/_utils.py` imports `FetchProperties` from `_action/_data.py`.
> 
> How do you feel about moving `get_timeout` to `_action/_data.py`, as that is the only place outside of testing where it is used?

That seems like a good solution.